test/lock.cc

```diff
@@ -347,34 +292,103 @@ TEST(LockTest, ConditionVariable) {
   k2.Start();
   k3.Start();
 
-  EXPECT_EQ(
-      std::future_status::timeout,
-      future1.wait_for(std::chrono::seconds(0)));
+  EXPECT_EQ(std::future_status::timeout,
+            future1.wait_for(std::chrono::seconds(0)));
 
-  EXPECT_EQ(
-      std::future_status::timeout,
-      future2.wait_for(std::chrono::seconds(0)));
+  EXPECT_EQ(std::future_status::timeout,
+            future2.wait_for(std::chrono::seconds(0)));
 
-  EXPECT_EQ(
-      std::future_status::timeout,
-      future3.wait_for(std::chrono::seconds(0)));
+  EXPECT_EQ(std::future_status::timeout,
+            future3.wait_for(std::chrono::seconds(0)));
 
   EXPECT_FALSE(*foo.NotifyFor(41));
 
   EXPECT_TRUE(*foo.NotifyFor(42));
 
   future1.get();
 
-  EXPECT_EQ(
-      std::future_status::timeout,
-      future2.wait_for(std::chrono::seconds(0)));
+  EXPECT_EQ(std::future_status::timeout,
+            future2.wait_for(std::chrono::seconds(0)));
 
-  EXPECT_EQ(
-      std::future_status::timeout,
-      future3.wait_for(std::chrono::seconds(0)));
+  EXPECT_EQ(std::future_status::timeout,
+            future3.wait_for(std::chrono::seconds(0)));
 
   EXPECT_TRUE(*foo.NotifyAllFor(42));
 
   future2.get();
   future3.get();
 }
+
+TEST(LockTest, ConditionVariableUseAfterFree) {
+  // A bug was caught in the wild where `ConditionVariable` would enqueue a
+  // waiting `eventual` for later notification, even if the waiting condition
+  // was already met at the time of creation. This would result in a
+  // use-after-free type bug where the `ConditionVariable` at a later point
+  // attempts to notify a stale eventual.
+  //
+  // The bug is demonstrated below using a simple binary condition object
+  // `Foo`, that can be either in the state `true` or `false`. Starting with
+  // the condition variable in the false state, an `eventual` is created (and
+  // started) that will simply wait for the `false` state and return the
+  // current state. This is non blocking and should not result in the
+  // `eventual` getting enqueued in the `ConditionVariable` for later
+  // notification. The eventual is then garbage collected.
+  //
+  // The state of the condition object is then "updated" to `false` and all the
+  // internal `waiter`s are notified. In case the destructed eventual is in the
+  // list the test will blow up with a message along the lines of
+  // "pure virtual method called" as the `ConditionVariable` attempts to
+  // `notify` the destructed object.
+
+  struct Foo : public Synchronizable {
+    Foo() : condition_variable_(&lock()) {}
+
+    auto SetCondition(bool value) {
+      return Synchronized(Then([this, value]() mutable {
+        condition_ = value;
+        condition_variable_.NotifyAll();
+        return value;
+      }));
+    }
+
+    auto WaitFor(bool value) {
+      return Synchronized(condition_variable_.Wait([this, value]() {
+        return condition_ != value;
+      }) | Just(value));
+    }
```

Comment:
I don't understand the value of the extra complexity in checking `condition_`? For the tests sake, why not just call `condition_variable_.Wait()` with a lambda that always returns `false`?

Likewise why not just simplify and rename `SetCondition` to something like `NotifyAll()` and drop the `bool condition_` member all together?

Then the test can then just do `*foo.Wait()` and then `*foo.NotifyAll()`.

There are other code paths here that are definitely not getting tested nor do I think we need/want to test and I'd like to not have to think about those when trying to understand what we are testing. 😉